gcp_variant_transforms/beam_io/vcf_header_io.py

```diff
@@ -0,0 +1,211 @@
+
+
+class _VcfHeaderSource(filebasedsource.FileBasedSource):
+  """A source for reading VCF file headers.
+
+  Parses VCF files (version 4) using PyVCF library.
+  """
+
+  DEFAULT_VCF_READ_BUFFER_SIZE = 65536  # 64 kB
+
+  class ReadBuffer(object):
+    # A buffer that gives the buffered data and next position in the
+    # buffer that should be read.
+
+    def __init__(self, data, position):
+      self._data = data
+      self._position = position
+
+    @property
+    def data(self):
+      return self._data
+
+    @data.setter
+    def data(self, value):
+      assert isinstance(value, bytes)
+      self._data = value
+
+    @property
+    def position(self):
+      return self._position
+
+    @position.setter
+    def position(self, value):
+      assert isinstance(value, (int, long))
+      if value > len(self._data):
+        raise ValueError('Cannot set position to %d since it\'s larger than '
+                         'size of data %d.', value, len(self._data))
+      self._position = value
+
+  def __init__(self,
+               file_pattern,
+               compression_type=CompressionTypes.AUTO,
+               buffer_size=DEFAULT_VCF_READ_BUFFER_SIZE,
+               validate=True):
+    super(_VcfHeaderSource, self).__init__(file_pattern,
+                                           compression_type=compression_type,
+                                           validate=validate,
+                                           splittable=False)
+    self._coder = coders.StrUtf8Coder()
+    self._buffer_size = buffer_size
+
+  def read_records(self, file_name, range_tracker):
+    try:
+      vcf_reader = vcf.Reader(fsock=self._read_headers(file_name))
+    except StopIteration:
+      raise ValueError('%s has no header.', file_name)
+
+    infos = vcf_reader.infos
+    formats = vcf_reader.formats
+    # These methods were not designed to be protected. They start with an
+    # underscore to avoid confilcts with field names. For more info, see below:
+    # https://docs.python.org/2/library/collections.html#collections.namedtuple
+    yield ({key: infos[key]._asdict() for key in infos},  # pylint: disable=W0212
+           {key: formats[key]._asdict() for key in formats})  # pylint: disable=W0212
+
+  def _read_headers(self, file_name):
+    read_buffer = _VcfHeaderSource.ReadBuffer('', 0)
+
+    with self.open_file(file_name) as file_to_read:
+      position = 0
+      while True:
+        record, num_bytes_to_next_record = self._read_record(file_to_read,
+                                                             read_buffer)
+        decoded_line = self._coder.decode(record)
+        if not decoded_line.startswith('#'):
+          break
+
+        yield decoded_line
+
+        if num_bytes_to_next_record < 0:
+          break
+        position += num_bytes_to_next_record
+
+  def _read_record(self, file_to_read, read_buffer):
+      # Returns a tuple containing the current_record and number of bytes to the
```

Comment:
Indentation is inconsistent on the comments here.

Reply:
Thanks! My editor must have added an indent when pasting.